File: iSeg/vtkAppendProjects.py
#!/usr/bin/env python
##
## Copyright (c) 2018 The Foundation for Research on Information Technologies in Society (IT'IS).
## 
## This file is part of iSEG
## (see https://github.com/ITISFoundation/osparc-iseg).
## 
## This software is released under the MIT License.
##  https://opensource.org/licenses/MIT
##
##




# Title: vtkAppendProjects
#
# Description: Append iSeg projects.
#
# Bugs:
#



import os,sys,string
from optparse import OptionParser
# import xml.dom.minidom as minidom
# import tables

from vtk import *
# from Xdmf import *
from libvtkXdmfPython import *
# from numpy import *
from libvtkmyFilteringPython import *
from libvtkmyImagingPython import *

usage = "usage: %prog [options] inputs output"
parser = OptionParser(usage=usage)
parser.add_option("--shift", type=int, default=0)
# parser.add_option("--background", type=int, default=0)
parser.add_option("--preserve", type=int, default=1)
parser.add_option("--axis", type=int, default=0)
# parser.add_option("--threads", type=int, default=1)

(opts, args) = parser.parse_args()
if len(args)<3:
	parser.error("incorrect number of arguments")

NINPUTS = len(args)-1
print "got", NINPUTS, "inputs"

# append = vtkImageAppend()
append = vtkImageAppend2()
# append.SetNumberOfThreads(1)
if opts.preserve:
	print "Preserving extents..."
	append.SetPreserveExtents(1)
else:
	print "Appending along axis..."
	append.SetPreserveExtents(0)
	append.SetAppendAxis(opts.axis)

append.SetUseTransparency(1)
# append.SetIntTransparentValue(opts.background)
# append.SetLongTransparentValueMin(0)
# append.SetLongTransparentValueMax(0)
# append.SetDoubleTransparentValueMin(-20.0)
# append.SetDoubleTransparentValueMax(20.0)

for i in range(NINPUTS):
	filename = args[i]
	print i, filename

	reader = vtkXdmfReader()
	reader.SetFileName(filename)
	reader.UpdateInformation()
	if not reader.IsFileStructuredPoints:
		print reader.GetClassName()
		exit(0)
	reader.SetPointArrayStatus("Source",1)
	reader.SetPointArrayStatus("Target",0)
	reader.SetPointArrayStatus("Tissue",1)
	reader.Update()

	# reader = vtkXMLImageDataReader()
	# reader.SetFileName(filename)
	# reader.Update()

	# info = vtkImageChangeInformation()
	# ext = reader.GetOutputDataObject(0).GetExtent()
	# info.Update()

	# dataset = reader.GetOutputDataObject(0)
	dataset = vtkImageData()
	dataset.DeepCopy(reader.GetOutputDataObject(0))
	# dataset.DeepCopy(reader.GetOutput())
	# dataset.DeepCopy(info.GetOutput())
	# dataset.GetPointData().SetActiveScalars("Source")


	org = dataset.GetOrigin()
	spc = dataset.GetSpacing()
	ext = dataset.GetExtent()

	print org
	print spc
	print ext

	if opts.shift:
		# The first dataset in the list becomes the reference.
		if i==0:
			org0 = org
			print "Reference origin:", org0
		else:
			dataset.SetOrigin(org0)
			print "new origin:", dataset.GetOrigin()
			dorg = [org[0]-org0[0], org[1]-org0[1], org[2]-org0[2]]
			dext = [int(dorg[0]/spc[0]), int(dorg[1]/spc[1]), int(dorg[2]/spc[2])]
			print "dext:", dext
			ext1 = [ext[0]+dext[0], ext[1]+dext[0], ext[2]+dext[1], ext[3]+dext[1], ext[4]+dext[2], ext[5]+dext[2]]
			print "new extent:", ext1
			dataset.SetExtent(ext1)

	# writer = vtkXMLImageDataWriter()
	# writer.SetInput(dataset)
	# writer.SetFileName("%s.vti" % filename)
	# writer.Write()

	dataset.SetScalarTypeToUnsignedChar()
	append.AddInput(dataset)



print "updating..."
append.Update()
output = append.GetOutput()
print output.GetOrigin()

"""
print "resampling..."
output.GetPointData().SetActiveScalars("Source")
output.SetScalarTypeToFloat()
resample = vtkImageReslice()
resample.SetInputArrayToProcess(0,0,0,0,"Source")
resample.SetInput(output)
resample.SetOutputSpacing(5,5,5)
resample.Update()

print "writing..."
writer = vtkXMLImageDataWriter()
# writer.SetInput(output)
writer.SetInput(resample.GetOutput())
writer.SetFileName("out-1.vti")
writer.Write()

output.GetPointData().SetActiveScalars("Tissue")
output.SetScalarTypeToUnsignedChar()
resample = vtkImageReslice()
resample.SetInputArrayToProcess(0,0,0,0,"Tissue")
resample.SetInput(output)
resample.SetOutputSpacing(5,5,5)
resample.Update()

print "writing..."
writer.SetInput(resample.GetOutput())
writer.SetFileName("out-2.vti")
writer.Write()
"""

print "writing..."
filename = args[NINPUTS]
print "writing output file", filename
writer = vtkXdmfWriter()
writer.SetInput(output)
writer.SetFileName(filename)
writer.Write()

print "END"

File: iSeg/vtkAppendProjects.cpp
/*
 * Copyright (c) 2018 The Foundation for Research on Information Technologies in Society (IT'IS).
 * 
 * This file is part of iSEG
 * (see https://github.com/ITISFoundation/osparc-iseg).
 * 
 * This software is released under the MIT License.
 *  https://opensource.org/licenses/MIT
 */

#include <cassert>
#include <set>
#include <string>

#include <vtkAppendImageData.h>
#include <vtkCell.h>
#include <vtkCellArray.h>
#include <vtkCellData.h>
#include <vtkDataSet.h>
#include <vtkDoubleArray.h>
#include <vtkIdList.h>
#include <vtkImageAppend2.h>
#include <vtkImageData.h>
#include <vtkIntArray.h>
#include <vtkPointData.h>
#include <vtkPoints.h>
#include <vtkSmartPointer.h>
#include <vtkVolume.h>
#include <vtkXMLImageDataReader.h>
#include <vtkXMLImageDataWriter.h>

#include <Options.h>
#include <Tools.h>

#define vtkNew(type, name)                                                     \
	vtkSmartPointer<type> name = vtkSmartPointer<type>::New()

using namespace std;
using namespace Tools;

int main(int argc, char **argv)
{
	string ifnames, ofname;
	string sname;
	int shift = 0;
	int background = 0;
	Options opts(argc, argv);
	opts.add("i", &ifnames);
	opts.add("o", &ofname);
	opts.add("scalar", &sname);
	opts.add("shift", &shift);
	opts.parse();

	opts.print();

	//	if(sname.empty())
	//		error("no scalar");

	vector<string> fnames = tokenize(ifnames, ',');
	if (fnames.size() < 2)
		error("need at least 2 inputs");
	if (ofname.empty())
		error("need output filename");

	//	vtkSmartPointer<vtkImageData> image;
	vtkNew(vtkXMLImageDataReader, reader);

	//	vtkNew(vtkAppendImageData,append);
	// if(shift) append->SetShiftByOrigin(1);
	//   vtkNew(vtkImageAppend,append);
	vtkNew(vtkImageAppend2, append);
	append->SetPreserveExtents(1);
	append->SetUseTransparency(1);

	// cerr << "using threads: " << append->GetNumberOfThreads() << endl;
	// append->SetNumberOfThreads(1);
	cerr << "using threads: " << append->GetNumberOfThreads() << endl;

	double org0[3] = {0.0, 0.0, 0.0};
	double org[3] = {0.0, 0.0, 0.0};
	double dorg[3] = {0.0, 0.0, 0.0};
	double spc[3] = {0.0, 0.0, 0.0};
	int ext[6] = {0, 0, 0, 0, 0, 0};
	int dext[6] = {0, 0, 0, 0, 0, 0};
	int ext1[6] = {0, 0, 0, 0, 0, 0};

	for (int i = 0; i < fnames.size(); ++i) {
		cerr << i << " " << fnames[i] << endl;
		reader->SetFileName(fnames[i].c_str());
		reader->Update();
		vtkNew(vtkImageData, image);
		image->DeepCopy(reader->GetOutput());
		//		append->AddInput(reader->GetOutput());
		// int code = image->GetPointData()->SetActiveScalars(sname.c_str());
		// if(code<0)
		// error("no such data: " + sname);

		image->GetOrigin(org);
		image->GetSpacing(spc);
		image->GetExtent(ext);

		if (shift) {
			if (i == 0) {
				memcpy(org0, org, 3 * sizeof(double));
			}
			else {
				image->SetOrigin(org0);
				cerr << "new origin: " << org0[0] << " " << org0[1] << " " << org0[2]
						 << endl;
				for (int j = 0; j < 3; ++j) {
					dorg[j] = org[j] - org0[j];
				}
				for (int j = 0; j < 3; ++j) {
					dext[j] = int(dorg[j] / spc[j]);
				}
				for (int j = 0; j < 3; ++j) {
					ext1[2 * j + 0] = ext[2 * j + 0] + dext[j];
					ext1[2 * j + 1] = ext[2 * j + 1] + dext[j];
				}
				cerr << "new extent: " << ext1[0] << " " << ext1[1] << " " << ext1[2]
						 << " " << ext1[3] << " " << ext1[4] << " " << ext1[5] << endl;
				image->SetExtent(ext1);
			}
		}

		vtkNew(vtkXMLImageDataWriter, writer);
		writer->SetDataModeToAppended();
		writer->SetEncodeAppendedData(0);
		writer->SetInput(image);
		char buf[1024];
		sprintf(buf, "test-%d.vti", i);
		writer->SetFileName(buf);
		writer->Write();

		append->AddInput(image);
	}

	append->Update();
	// append->GetOutput()->GetPointData()->GetScalars()->SetName(sname.c_str());
	cerr << "writing output image..." << endl;

	//	string obname, oextname;
	//	Tools::decomposeFileName(obname,oextname,ofname);
	vtkNew(vtkXMLImageDataWriter, writer);
	writer->SetDataModeToAppended();
	writer->SetEncodeAppendedData(0);
	writer->SetInput(append->GetOutput());
	writer->SetFileName(ofname.c_str());
	writer->Write();

	return EXIT_SUCCESS;
}
